src/tools/rbd_mirror/image_map/Types.h

```diff
@@ -0,0 +1,116 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <set>
+#include <map>
+#include <string>
+#include <memory>
+
+#include "cls/rbd/cls_rbd_types.h"
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum ShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct Image {
+  // to_instance_id is unused as of now and always
+  // matches instance_id. When we have rebalancer
+  // in place to_instance_id would point to the
+  // instance id of the mirror daemon where the
+  // image needs to be reassigned (from currently
+  // instance_id).
+  std::string id;
+  std::string global_image_id;
+  std::string instance_id;
+  std::string to_instance_id;
+
+  Image(const std::string &global_image_id)
+    : global_image_id(global_image_id) {
+  }
+
+  Image(const std::string &id,
+        const std::string &global_image_id)
+    : id(id), global_image_id(global_image_id) {
+  }
+
+  Image(const std::string &id,
+        const std::string &global_image_id,
+        const std::string &instance_id)
+    : id(id),
+      global_image_id(global_image_id),
+      instance_id(instance_id),
+      to_instance_id(instance_id) {
+  }
+
+  Image(const std::string &id,
+        const std::string &global_image_id,
+        const std::string &instance_id,
+        const std::string &to_instance_id)
+    : id(id),
+      global_image_id(global_image_id),
+      instance_id(instance_id),
+      to_instance_id(to_instance_id) {
+  }
+
+  inline bool operator==(const Image &rhs) const {
+    return (global_image_id == rhs.global_image_id && id == rhs.id);
+  }
+
+  inline bool operator<(const Image &rhs) const {
+    return global_image_id < rhs.global_image_id;
+  }
+};
+
+std::ostream &operator<<(std::ostream &, const Image &image);
+
+typedef std::set<Image> Images;
+
+struct MapSpec {
+  Images added;
+  Images removed;
+
+  MapSpec() {
+  }
+
+  MapSpec(const Images &added,
+          const Images &removed)
+    : added(added),
+      removed(removed) {
+  }
+};
+
+std::ostream &operator<<(std::ostream &, const MapSpec &mapspec);
+
+// dynamically built information about an image
+struct ImageInfo {
+  // TODO: track state once we maintain image state
+  // on-disk
+  ImageInfo() {
+  }
+
+  std::set<std::string> peer_uuid;
+};
+
+std::ostream &operator<<(std::ostream &, const ImageInfo &info);
+
+typedef std::set<std::string> GlobalImageIds;
+typedef std::map<std::string, GlobalImageIds> InMap;
```

Comment:
I have a real irrational hatred of this name. I think this essentially should be a private typedef of ```Policy``` since the ```LoadRequest``` only really needs to pass a map of global image ids -> instance ids (just like the combined ```UpdateRequest``` would just take a map of global image ids -> instance ids for update / removal (empty instance id could reflect removal).